### PR TITLE
Bump Pkg

### PR DESCRIPTION
**Pull request by KristofferC**
* Now makes it possible to declare compatibility requirements on packages in the Project. Some docs included. Please try it out.
* Fix some bug when testing projects, thanks @staticfloat for testing things out.
* Some other minor tweaks and small bugfixes.

### CODE REVIEW

**Comment by staticfloat** [comment]
On OSX:

```
Some tests did not pass: 2 passed, 0 failed, 1 errored, 0 broken.OldPkg/pkg: Error During Test at /private/tmp/julia/share/julia/test/testdefs.jl:19
  Got exception LoadError("/private/tmp/julia/share/julia/stdlib/v0.7/OldPkg/test/pkg.jl", 56, KeyError("SparseArrays")) outside of a @test
  LoadError: KeyError: key "SparseArrays" not found
  Stacktrace:
   [1] getindex at ./dict.jl:487 [inlined]
   [2] prune_versions(::Dict{String,OldPkg.Types.VersionSet}, ::Dict{String,Dict{VersionNumber,OldPkg.Types.Available}}, ::Dict{AbstractString,OldPkg.Types.ResolveBacktraceItem}) at /Users/travis/build/JuliaLang/julia/usr/share/julia/stdlib/v0.7/OldPkg/src/query.jl:421
   [3] prune_dependencies at /Users/travis/build/JuliaLang/julia/usr/share/julia/stdlib/v0.7/OldPkg/src/query.jl:586 [inlined]
   [4] resolve(::Dict{String,OldPkg.Types.VersionSet}, ::Dict{String,Dict{VersionNumber,OldPkg.Types.Available}}, ::Dict{String,Tuple{VersionNumber,Bool}}, ::Dict{String,OldPkg.Types.Fixed}, ::Dict{String,VersionNumber}, ::Set{String}) at /Users/travis/build/JuliaLang/julia/usr/share/julia/stdlib/v0.7/OldPkg/src/entry.jl:508
   [5] resolve at /Users/travis/build/JuliaLang/julia/usr/share/julia/stdlib/v0.7/OldPkg/src/entry.jl:487 [inlined] (repeats 6 times)
   [6] #3 at /Users/travis/build/JuliaLang/julia/usr/share/julia/stdlib/v0.7/OldPkg/src/dir.jl:33 [inlined]
   [7] cd(::getfield(OldPkg.Dir, Symbol("##3#6")){Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},typeof(OldPkg.Entry.resolve),Tuple{}}, ::String) at ./file.jl:96
   [8] #2 at /Users/travis/build/JuliaLang/julia/usr/share/julia/stdlib/v0.7/OldPkg/src/dir.jl:33 [inlined]
   [9] withenv(::getfield(OldPkg.Dir, Symbol("##2#5")){Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}},typeof(OldPkg.Entry.resolve),Tuple{},String}, ::Pair{String,String}) at ./env.jl:148
   [10] #cd#1(::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}, ::Function, ::Function) at /Users/travis/build/JuliaLang/julia/usr/share/julia/stdlib/v0.7/OldPkg/src/dir.jl:32
   [11] cd at /Users/travis/build/JuliaLang/julia/usr/share/julia/stdlib/v0.7/OldPkg/src/dir.jl:25 [inlined]
   [12] resolve at /Users/travis/build/JuliaLang/julia/usr/share/julia/stdlib/v0.7/OldPkg/src/OldPkg.jl:233 [inlined]
   [13] (::getfield(Main.Test23Main_OldPkg_pkg.PkgTests, Symbol("##6#7")){Bool,getfield(Main.Test23Main_OldPkg_pkg.PkgTests, Symbol("##8#55")),String,Bool})() at /private/tmp/julia/share/julia/stdlib/v0.7/OldPkg/test/pkg.jl:37
   [14] withenv(::getfield(Main.Test23Main_OldPkg_pkg.PkgTests, Symbol("##6#7")){Bool,getfield(Main.Test23Main_OldPkg_pkg.PkgTests, Symbol("##8#55")),String,Bool}, ::Pair{String,String}) at ./env.jl:148
   [15] #temp_pkg_dir#5(::Bool, ::Function, ::getfield(Main.Test23Main_OldPkg_pkg.PkgTests, Symbol("##8#55")), ::String, ::Bool) at /private/tmp/julia/share/julia/stdlib/v0.7/OldPkg/test/pkg.jl:31
   [16] temp_pkg_dir(::Function, ::String, ::Bool) at /private/tmp/julia/share/julia/stdlib/v0.7/OldPkg/test/pkg.jl:31 (repeats 2 times)
   [17] top-level scope
   [18] include at ./boot.jl:314 [inlined]
   [19] include_relative(::Module, ::String) at ./loading.jl:1071
   [20] include(::Module, ::String) at ./sysimg.jl:29
   [21] include(::String) at /private/tmp/julia/share/julia/test/testdefs.jl:13
   [22] macro expansion at /private/tmp/julia/share/julia/test/testdefs.jl:22 [inlined]
   [23] macro expansion at /Users/travis/build/JuliaLang/julia/usr/share/julia/stdlib/v0.7/Test/src/Test.jl:1080 [inlined]
   [24] macro expansion at /private/tmp/julia/share/julia/test/testdefs.jl:21 [inlined]
   [25] macro expansion at ./util.jl:289 [inlined]
   [26] top-level scope at ./<missing>:19
   [27] eval at ./boot.jl:316 [inlined]
   [28] #runtests#3(::UInt128, ::Function, ::String, ::String, ::Bool) at /private/tmp/julia/share/julia/test/testdefs.jl:25
   [29] #runtests at ./<missing>:0 [inlined] (repeats 2 times)
   [30] (::getfield(Distributed, Symbol("##112#114")){Distributed.CallMsg{:call_fetch}})() at /Users/travis/build/JuliaLang/julia/usr/share/julia/stdlib/v0.7/Distributed/src/process_messages.jl:269
   [31] run_work_thunk(::getfield(Distributed, Symbol("##112#114")){Distributed.CallMsg{:call_fetch}}, ::Bool) at /Users/travis/build/JuliaLang/julia/usr/share/julia/stdlib/v0.7/Distributed/src/process_messages.jl:56
   [32] macro expansion at /Users/travis/build/JuliaLang/julia/usr/share/julia/stdlib/v0.7/Distributed/src/process_messages.jl:269 [inlined]
   [33] (::getfield(Distributed, Symbol("##111#113")){Distributed.CallMsg{:call_fetch},Distributed.MsgHeader,Sockets.TCPSocket})() at ./task.jl:257
  in expression starting at /private/tmp/julia/share/julia/stdlib/v0.7/OldPkg/test/pkg.jl:56
```

**Comment by KristofferC** [comment]
That's odd but seems to be an OldPkg test and should be unrelated. Added another commit so CI will rerun.

**Comment by staticfloat** [comment]
I've restarted some builds now  that METADATA.jl has been cleaned up, which should fix the `SparseArrays` problems.

File: stdlib/Pkg/docs/src/index.md [hunk 1]
```diff
@@ -394,18 +394,73 @@ or
...
+
+!!! info
+  There is currently no way to give compatibility from the Pkg REPL mode so for now, one has to manually edit the project file.
+
+### Version specifier format
+
+Similar to other package managers, the Julia package manager respects [semantic versioning](https://semver.org/) (semver).
+As an example, a version specifier is given as e.g. `1.2.3` is therefore assumed to be compatible with the versions `[1.2.3 - 2.0.0)` where `)` is a non-inclusive upper bound.
+More specifically, a version specifier is either given as a **caret specifier**, e.g. `~1.2.3`  or a **tilde specifier** `^1.2.3`.
+Caret specifiers are the default and hence `1.2.3 == ~1.2.3`. The difference between a caret and tilde is described in the next section.
```
Comment (GunnarFarneback) [inline stdlib/Pkg/docs/src/index.md]: Also looks like a caret and tilde mix up.

File: stdlib/Pkg/docs/src/index.md [hunk 2]
```diff
@@ -394,18 +394,73 @@ or
...
+^0.0 = [0.0.0, 0.1.0)
+^0 = [0.0.0, 1.0.0)
+```
+
+While the semver specification says that all versions with a major version of 0 are incompatible with each other, we have made that choice that
+a version given as `0.a.b` is considered compatible with `0.a.c` if `a != 0` and  `c >= b`.
+
+#### Tilde specifiers
+
+A tilde specifier provides more limited upgrade possibilities. With a caret, only the last specified digit is allowed to increment by one.
```
Comment (GunnarFarneback) [inline stdlib/Pkg/docs/src/index.md]: Should say "tilde" both times?

**Comment by StefanKarpinski** [comment]
@GunnarFarneback: thanks for the review, see also https://github.com/JuliaLang/Pkg.jl/pull/348 which makes some related documentation fixes. I think we'll have to wait for @KristofferC to be back next week to tell us which is correct.

**Comment by staticfloat** [comment]
@KristofferC Woohoo!  Confirmed that Pkg3 "just works" with `BinaryProvider` now, with the default Travis test script.  :)

https://travis-ci.org/JuliaPackaging/BinaryProvider.jl/jobs/391963313

Note that OSX hasn't picked up these changes yet, but we've got new versions coming down the pipe, so it shouldn't be too long now.  Thanks!

**Pull request merged**